### PR TITLE
Only push output device messages to Test Explorer, don't push logs

### LINKED ISSUE #4213: Output messages should be forward to clients instead of logs (server mode)
As of today, the logs are forwarded to the client and are displayed there. This is not right because the logs are meant to be troubleshooting oriented hence untranslated and possibly with many messages intended to help platform/framework/extension authors and not directly the end user.

On the other hand, the output devices messages are designed as user facing messaging and are currently not forwarded to be printed into the client (e.g. Test Explorer).

This is leading to multiple localization issues and missing some important/useful notifications for the end users.

### PR DESCRIPTION
**Pull request by Youssef1313**
Logs are intended to be used for diagnosing. They shouldn't be localized and they shouldn't be pushed to Test Explorer.

Fixes #4213

Linked to AB#2275339

### CODE REVIEW

File: test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ServerModeTests.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ public async Task DiscoverAndRun(string tfm)
         await Task.WhenAll(discoveryListener.WaitCompletion(), runListener.WaitCompletion());
         Assert.AreEqual(1, discoveryCollector.TestNodeUpdates.Count(x => x.Node.NodeType == "action"), $"Wrong number of discovery");
         Assert.AreEqual(2, runCollector.TestNodeUpdates.Count, $"Wrong number of updates");
-        Assert.IsFalse(logs.IsEmpty, $"Logs are empty");
+        Assert.IsTrue(logs.IsEmpty, $"Logs are not empty");
```
Comment (Youssef1313) [inline test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ServerModeTests.cs]: Need to carefully double check if this is expected. And if yes, it's ideal to have another test that shows logs being not empty.

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs [hunk 2]
```diff
@@ -19,11 +34,72 @@ internal class ServerModePerCallOutputDevice : IPlatformOutputDevice
 
     public Task DisplayAfterSessionEndRunAsync() => Task.CompletedTask;
 
-    public Task DisplayAsync(IOutputDeviceDataProducer producer, IOutputDeviceData data) => Task.CompletedTask;
+    public async Task DisplayAsync(IOutputDeviceDataProducer producer, IOutputDeviceData data)
+    {
+        using (await _asyncMonitor.LockAsync(TimeoutHelper.DefaultHangTimeSpanTimeout))
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs]: I don't think we need this lock, we could just forward down and the thing below is responsible for that.

@MarcoRossignoli shall we lock to ensure ordering below?

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs [hunk 3]
```diff
@@ -19,11 +34,72 @@ internal class ServerModePerCallOutputDevice : IPlatformOutputDevice
...
+                    break;
+
+                case ExceptionOutputDeviceData exceptionOutputDeviceData:
+                    await LogAsync(LogLevel.Error, exceptionOutputDeviceData.Exception.ToString(), padding: null);
+                    break;
+            }
+        }
+    }
 
     public Task DisplayBannerAsync(string? bannerMessage) => Task.CompletedTask;
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs]: Let's handle this case and print as debug.

@MarcoRossignoli @drognanar it feels like it would be too verbose to print as Info for TE but maybe we still prefer doing it. WDYT?

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs [hunk 4]
```diff
@@ -19,11 +34,72 @@ internal class ServerModePerCallOutputDevice : IPlatformOutputDevice
...
 
     public Task DisplayBeforeSessionStartAsync() => Task.CompletedTask;
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs]: Let's emit a debug message here and when in trace we should emit a localized message with the path to the log file to help users looking up the log.

**Review by Evangelink — COMMENTED** [review]
Let's add a test where we have a fake adapter that emits the various messages and check we catch them with the right level.

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs [hunk 5]
```diff
@@ -450,13 +444,20 @@ private async Task<ResponseArgsBase> ExecuteRequestAsync(RequestArgsBase args, s
 
         DateTimeOffset adapterLoadStart = _clock.UtcNow;
 
+        IPlatformOutputDevice outputDevice = ServiceProvider.GetRequiredService<IPlatformOutputDevice>();
+        // TODO: What if some one wants to have their own output device that's also compatible with server mode in IDE?
```
Comment (MarcoRossignoli) [inline src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs]: I think here we need to do something better. 

Despite `IPlatformOutputDevice` is a bit special and the implementation needs to take into  account more scenario(like app restart and correct handling for the "output") I'd like to abstract it from the "modes". 

`IPlatformOutputDevice` implements `IOutputDevice` and today we pass to the `IOutputDevice` requestors the `IPlatformOutputDevice` implementation as-is. 

Clearly this is not more good if we need to route display info due to modes.

So we should when  `IOutputDevice` is requested from an extension to proxy it to be able to forward to the `IPlatformOutputDevice` but also to other internal part of the platform that needs it, for instance the server mode.

High level idea
```
 GetService<IOutputDevice>() -> OutputDeviceRouter -> ArbitraryPlatformOutputDeviceImplementation & ServerModeForwarder
```

In this way we don't need special or custom output device implementation for old and possible new modes and we can keep the "current" output device choice.

We cannot forward the `ArbitraryPlatformOutputDeviceImplementation` display..but I would expect that the info printed by the "device" are not important...the device output goal is to "show information" in "some way" it doesn't affect and should not affect in any way the execution, so information produced by this extension shouldn't be so useful.

cc: @Evangelink

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ProxyPlatformOutputDevice.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+using Microsoft.Testing.Platform.Helpers;
+using Microsoft.Testing.Platform.Hosts;
+using Microsoft.Testing.Platform.OutputDevice;
+using Microsoft.Testing.Platform.TestHost;
+
+namespace Microsoft.Testing.Platform.ServerMode;
+
+// Any interfaces that can have special treatment for any output device should be implemented by
+// this proxy class and be forwarded properly.
+// This is not so good. How can we make sure we are not missing any interfaces that may be implemented by external output devices?
```
Comment (Youssef1313) [inline src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ProxyPlatformOutputDevice.cs]: @Evangelink @MarcoRossignoli Can you think of a good way to solve this?

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs [hunk 7]
```diff
@@ -93,13 +96,38 @@ public ServerTestHost(
...
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+    private async void OnCurrentDomainUnhandledException(object sender, UnhandledExceptionEventArgs e)
+#pragma warning restore VSTHRD100 // Avoid async void methods
+    {
+        _logger.LogWarning($"[ServerTestHost.OnCurrentDomainUnhandledException] {e.ExceptionObject}{_environment.NewLine}IsTerminating: {e.IsTerminating}");
+
+        // Looks like nothing in this message to really be localized?
+        // All are class names, method names, property names, and placeholders. So none is localizable?
+        await ServiceProvider.GetOutputDevice().DisplayAsync(
```
Comment (MarcoRossignoli) [inline src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs]: I think that use the `GetAwaiter().GetResult()` is better here. I don't expect a log of call here and awaiting a void we could lose the message.

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/OutputDeviceManager.cs [hunk 8]
```diff
@@ -15,15 +16,15 @@ public void SetPlatformOutputDevice(Func<IServiceProvider, IPlatformOutputDevice
         _platformOutputDeviceFactory = platformOutputDeviceFactory;
     }
 
-    public IPlatformOutputDevice Build(ServiceProvider serviceProvider)
+    internal ProxyPlatformOutputDevice Build(ServiceProvider serviceProvider, bool useServerModeOutputDevice)
```
Comment (MarcoRossignoli) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/OutputDeviceManager.cs]: why internal ctor?

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ProxyPlatformOutputDevice.cs (hunk 6 above)
Comment (MarcoRossignoli) [reply src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ProxyPlatformOutputDevice.cs]: Why do we need it?

The proxy should only implement `IOutputDevice` and returned when users asks to the service provider. Here after the proxy should have today at max 2 "receivers", the real `IPlatformOutputDevice` and the servicehost logger(you can directly add a method to that instead to implement a custom one).

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs [hunk 9]
```diff
@@ -93,13 +96,36 @@ public ServerTestHost(
...
+    {
+        _logger.LogWarning($"[ServerTestHost.OnCurrentDomainUnhandledException] {e.ExceptionObject}{_environment.NewLine}IsTerminating: {e.IsTerminating}");
+
+        // Looks like nothing in this message to really be localized?
+        // All are class names, method names, property names, and placeholders. So none is localizable?
+        ServiceProvider.GetOutputDevice().DisplayAsync(
+            this,
+            new WarningMessageOutputDeviceData(
+                $"[ServerTestHost.OnCurrentDomainUnhandledException] {e.ExceptionObject}{_environment.NewLine}IsTerminating: {e.IsTerminating}"))
+            .GetAwaiter().GetResult();
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs:118]: @MarcoRossignoli is this better to do this or use await on async void? I think I recall the await async void was the "recommended" way for eventing methods in WPF but I might say something wrong.

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs [hunk 10]
```diff
@@ -269,6 +288,12 @@ private async Task HandleNotificationAsync(NotificationMessage message, Cancella
                         if (cancellationException is null)
                         {
                             await _logger.LogWarningAsync($"Exception during the cancellation of request id '{args.CancelRequestId}'");
+
+                            // TODO: Localize
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs]: TODO to fix

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs (hunk 9 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs:118]: @Evangelink I initially had it `async void` as I'm coming from the background that in general async void is okay for events in UI frameworks (e.g, WPF). Marco suggested using `GetAwaiter().GetResult()` instead.

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/ProxyOutputDevice.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+using Microsoft.Testing.Platform.Extensions.OutputDevice;
+using Microsoft.Testing.Platform.Hosts;
+using Microsoft.Testing.Platform.ServerMode;
+
+namespace Microsoft.Testing.Platform.OutputDevice;
+
+// Any interfaces that can have special treatment for any output device should be implemented by
+// this proxy class and be forwarded properly.
+// This is not so good. How can we make sure we are not missing any interfaces that may be implemented by external output devices?
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/ProxyOutputDevice.cs]: Comment seems to be outdated and could be dropped.

Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/ProxyOutputDevice.cs]: It is. Thanks!

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs [hunk 12]
```diff
@@ -25,6 +25,7 @@ namespace Microsoft.Testing.Platform.OutputDevice;
 /// <summary>
 /// Implementation of output device that writes to terminal with progress and optionally with ANSI.
 /// </summary>
+// NOTE: !!! Any interfaces here will likely need to be implemented by ProxyPlatformOutputDevice as well. !!!
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs]: Outdated?

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs [hunk 13]
```diff
@@ -1,14 +1,44 @@
...
+    private readonly ConcurrentBag<ServerLogMessage> _messages = new();
+
+    private IServerTestHost? _serverTestHost;
+
+    private static readonly string[] NewLineStrings = { "\r\n", "\n" };
+
+    public ServerModePerCallOutputDevice(IServiceProvider serviceProvider)
+        => _serviceProvider = serviceProvider;
+
+    internal async Task InitializeAsync(IServerTestHost serverTestHost)
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs:30]: Could you add a small comment for future readers as to why we do that

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj [hunk 14]
```diff
@@ -1,17 +1,21 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <TargetFrameworks>$(NetCurrent)</TargetFrameworks>
     <TestRunnerAdditionalArguments>$(TestRunnerAdditionalArguments) --retry-failed-tests 3</TestRunnerAdditionalArguments>
     <UseVSTestRunner>false</UseVSTestRunner>
     <DefineConstants Condition=" '$(FastAcceptanceTest)' == 'true'">$(DefineConstants);SKIP_INTERMEDIATE_TARGET_FRAMEWORKS</DefineConstants>
+    <DefineConstants>$(DefineConstants);IS_MTP</DefineConstants>
```
Comment (Evangelink) [inline test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj]: Unused?

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj [hunk 15]
```diff
@@ -1,17 +1,21 @@
...
     <OutputType>Exe</OutputType>
     <UseAppHost>true</UseAppHost>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="MSBuild.StructuredLogger" />
+    <PackageReference Include="StreamJsonRpc" />
     <PackageReference Include="Microsoft.Testing.Extensions.Retry" />
+
+    <Compile Include="$(RepoRoot)test\IntegrationTests\MSTest.Acceptance.IntegrationTests\ServerMode\**\*.cs" Link="%(RecursiveDir)%(FileName)%(Extension)" />
```
Comment (Evangelink) [inline test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj]: Nit: I would reverse the order (e.g. move file here and reference from MSTest) as platform is more "core" than MSTest but I am fine if we keep as-is.

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs (hunk 12 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs]: Yup. Will drop it

**Review by Evangelink — APPROVED** [review]
Some nits and questions but LGTM.

Thanks for the massive work here @Youssef1313! And thanks @MarcoRossignoli for technical consulting <3

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj (hunk 15 above)
Comment (Youssef1313) [reply test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj]: Let's do it as follow-up to not pollute the diff with too many file moves.

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs (hunk 9 above)
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs:118]: Ok so let's go with this suggestion. Thanks!

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj (hunk 15 above)
Comment (Youssef1313) [reply test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj]: Tracking in https://github.com/microsoft/testfx/issues/4231

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj (hunk 14 above)
Comment (Youssef1313) [reply test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests.csproj]: Removed

File: src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs (hunk 13 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/ServerMode/JsonRpc/ServerModePerCallOutputDevice.cs:30]: Added a comment. Does it look good?

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs (hunk 12 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/TerminalOutputDevice.cs]: Fixed

File: src/Platform/Microsoft.Testing.Platform/OutputDevice/ProxyOutputDevice.cs (hunk 11 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/OutputDevice/ProxyOutputDevice.cs]: Fixed

File: src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs (hunk 10 above)
Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/Hosts/ServerTestHost.cs]: Done

**Review by Evangelink — APPROVED** [review]
